### DISCUSSION TITLE
RTU Server & client on same RS485

### DISCUSSION THREAD

**Discussion by robotiko** [eModbus/eModbus#324] [Q&A] [answered]
Somehow related to  #261 ,  #257 , #270

Have an scenario where we have RTU server  and then act as a clients write message to several RTU Servers. 
A very simplified concept of bridge on the same Serial port. 
The main application is a simple server, but need to broadcast from time to time some messages received. 

Considering very reduced client needs.  Is is posible to use the  begin end of Server and client to reuse the hardware port? 

The idea would be: 
1. Have the server working, when a special 06 register, get the value and attend request
2. Disable server ->  Enable client  -> send broadcast -> disable client -> enable Server

trying, this flow, when disable server and enable client, I loose node comm, maybe related to improper serial establishing

Sketch 
void bridgeMessage(uint16_t address, int value){

    MBserver.end();//Dissable working server on  Serial485 to free serial port
    MB.begin(Serial485); //enable client on freed serial port. 
   
   //.. message broadcast section 
    
    MB.end(); //Disable client to free serial porta again
    MBserver.begin(Serial485); //reenable Server
}

Should this work?

**Comment by Miq1** [comment, marked as answer]
Uh... This concept would be violating some basic assumptions for Modbus RTU: there can be only one client and only clients may broadcast, so a client will not listen to broadcasts at all.

I take it you need to publish some server's data asynchronically to other devices on the bus. If most listeners will not need to individually request data, but only the published, those could be servers themselves listening on broadcasts only. A single dedicated client would have to periodically request the server holding the data and in turn broadcast these. As a consequence, only the single client is able to request, like the standard mandates.

**Reply by robotiko** [reply]
Ups  sorry, there was a typ in server /client  .. I will edit original question for the records.

Yes, just Servers receive broadcasts. 

Yes, we have serveral Servers and one will temporary also act as Client sending broadcast ("Simplified Bridge") to the rest. 
Just server listen to broadcasts. 

The scenario is  the "bridge" (Server) will receive request from external RTU Client and will write on several actuator on bus (Servers)  (broadcast). 


So the same port reuse can be feasible?  the reason why it gets stuck cna be the port init in client?
Standard.

**Reply by Miq1** [reply]
To start from the end, yes, the init processing probably will be the issue.

But again: it reads like you may consider using a RTU/RTU bridge and two RTU buses. One will have all the actuators listening, the bridge being the client, the other will have the bridge as a server, that is requested from the external client. 

The bridge of course will have to use two Serial ports to connect to both buses, but f.i. the buses even do not need to run with the same speed. The bridge may act as a server as well for local data with a dedicated ID for that.

**Reply by robotiko** [reply]
Hi Miq1, 
the main issue is that we are using T-can485 very nice but they have just 1 RS485 and Canbus. 
Just need to broadcast from time to time register written at server so other Servers can sync, not a full RTU client. 

I was also looking at the option so use vanilla  RTUutils::send using the RTU Server serial port already in use.

**Reply by Miq1** [reply]
The T-CAN485 has free GPIO connectors you can use for UART - with a RS485 adapter you will be able to drive a second bus. What you are trying to achieve will not work on a single bus.

**Reply by robotiko** [reply]
Thanks for the support and feedback @Miq1
